blog: 0.22 performance improvements (#2385)
kestra-io/docs: content/blogs/2025-04-08-performance-improvements.md

We had feedback from users processing a lot of executions per day that this `queues` table can grow big (in tens or even hundreds of gibabytes) and sometimes induce a very high load on the database.

We decided to improve our `queues` table cleaning in two ways:
- Purging all execution related messages at the end of the execution, keeping only the last execution message so late consumers can still be updated with the execution terminal state.
- Purging earlier than the default of 7 days messages that are known to be of high cardinality (logs, metrics and audit logs) but are only consumed by a single consumer, so we are sure that if it has been processed, we can clean them. They are now cleaned after 1 hour instead of 7 days.

With these two changes, on contrived benchmarks, the number of records of the `queues` table was reduced to 95%!

For more details, you can have a look at these two pull requests:
- https://github.com/kestra-io/kestra/pull/7286
- https://github.com/kestra-io/kestra/pull/7363

## Sneak peak of 0.22 versus 0.21

> We plan to go deeper into the performance of Kestra in a later blog post, but here is a performance comparison of 0.22 versus 0.21. What's important here are not the raw numbers but the difference between the two sets.

Reviewer: ```suggestion
We plan to discuss Kestra's performance further in a later blog post, but here is a performance comparison of 0.22 versus 0.21. What's important here is not the raw numbers but the difference between the two sets.
```

> The benchmark scenario is a flow triggered by a Kafka Realtime Trigger that, for each message, do a JSON transformation and return the output in a second task.

Reviewer: ```suggestion
The benchmark scenario is a flow triggered by a Kafka Realtime Trigger that performs a JSON transformation for each message and returns the output in a second task.
```

> We will generate 1000 executions by publishing messages to a Kafka topic at 10, 25, 50,75 and 100 messages per second, then check the latency of the execution by looking at the last execution of the scenario run.

Reviewer: ```suggestion
We generate 1000 executions by publishing messages to a Kafka topic at 10, 25, 50, 75, and 100 messages per second, then check the execution latency by looking at the last execution of the scenario run.
```

> 

Reviewer: ```suggestion

::collapse{title="Expand to see the Benchmark Flow"}
```

```yaml
id: realtime-kafka-json
namespace: company.team

triggers:
  - id: kafka-logs
    type: io.kestra.plugin.kafka.RealtimeTrigger
    topic: test_kestra
    properties:
      bootstrap.servers: localhost:9092
    groupId: myGroup

tasks:
  - id: transform
    type: io.kestra.plugin.transform.jsonata.TransformValue
    from:  "{{trigger.value}}"
    expression: |
      {
        "order_id": order_id,
        "customer_name": first_name & ' ' & last_name,
        "address": address.city & ', ' & address.country,
        "total_price": $sum(items.(quantity * price_per_unit))
      }
  - id: hello
    type: io.kestra.plugin.core.output.OutputValues
    values:
      log: "{{outputs.transform.value}}"
```

> 

Reviewer: ```suggestion
::
```

> TODO hide the flow by default

Reviewer: ```suggestion
```

### JDBC backend

| JDBC 0.21 - throughput  | JDBC 0.21 - latency | JDBC 0.22 - throughput | JDBC 0.22 - latency |
|-------------------------|---------------------|------------------------|---------------------|
| 10 exec/s               | 400ms               | 10 exec/s              | 150ms               |
| 25 exec/s               | 26s                 | 25 exec/s              | 2.6s                |
| 50 exec/s               | 43s                 | 50 exec/s              | 15s                 |
| 75 exec/s               | 49s                 | 75 exec/s              | 22s                 |
| 100 exec/s              | 59s                 | 100 exec/s             | 25s                 |

**Key takeaways**:

> - Performance has been greatly improved in 0.22, even when executions are not run concurrently (which is almost the case at 10 executions/s).

Reviewer: ```suggestion
- Performance has improved dramatically in 0.22, even when executions are not run concurrently (which is almost the case at 10 executions/s).
```

> - At high throughput performance starts to degrade vastly around 50 executions/s.

Reviewer: ```suggestion
- Performance starts to degrade vastly around throughput of 50 executions/s.
```

### Kafka backend

| Kafka 0.21 - throughput | Kafka 0.21 - latency | Kafka 0.22 - throughput | Kafka 0.22 - latency |
|-------------------------|----------------------|-------------------------|----------------------|
| 10 exec/s               | 800ms                | 10 exec/s               | 200ms                |
| 25 exec/s               | 800ms                | 25 exec/s               | 225ms                |
| 50 exec/s               | 900ms                | 50 exec/s               | 225ms                |
| 75 exec/s               | 1s                   | 75 exec/s               | 300ms                |
| 100 exec/s              | 1s                   | 100 exec/s              | 300ms                |
| 150 exec/s              | 1.2s                 | 150 exec/s              | 750ms                |
| 200 exec/s              | 2s                   | 200 exec/s              | 1.9s                 |

**Key takeaways**:

> - In 0.21, our Kafka backend is capable of sustaining higher throughput than our JDBC backend, but on low throughput, latency is more than the JDBC backend.

Reviewer: ```suggestion
- In 0.21, our Kafka backend can sustain higher throughput than our JDBC backend, but on low throughput, latency is more than the JDBC backend.
```

> - In 0.22, our Kafka backend achieves almost the same latency at low throughput than our JDBC backend. At up to 100 executions per second, latency didn't increase much, and in all cases it stayed under the latency achieved in 0.21.

Reviewer: ```suggestion
- In 0.22, our Kafka backend achieves almost the same latency at low throughput as our JDBC backend. At up to 100 executions per second, latency didn't increase much, and in all cases, it stayed under the latency, which was achieved in 0.21.
```
